cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactRepository.java

```diff
@@ -781,4 +790,102 @@ private void writeSystemMetadata(Id.Artifact artifactId, ArtifactInfo artifactIn
+   * @param methodName name of the method
+   * @param parameters request body which corresponds to method parameter
+   * @return Serialized string response from method invocation
+   * @throws ArtifactNotFoundException
+   * @throws IOException
+   * @throws IllegalArgumentException
+   * @throws BadRequestException
+   * @throws ClassNotFoundException
+   * @throws InvocationTargetException
+   */
+  public String pickParentArtifactAndInvokePluginEndpoint(Id.Namespace namespace,
+                                                          Id.Artifact artifactId, String pluginType,
+                                                          String pluginName, String methodName,
+                                                          String parameters) throws ArtifactNotFoundException,
+    IOException, IllegalArgumentException, BadRequestException, ClassNotFoundException, InvocationTargetException {
+    ArtifactDetail artifactDetail = artifactStore.getArtifact(artifactId);
+    return invokePluginEndpoint(namespace, artifactDetail, pluginType, pluginName,
+                                pickParentArtifact(artifactDetail), methodName, parameters);
+  }
+
+  private ArtifactDescriptor pickParentArtifact(ArtifactDetail artifactDetail)
+    throws InvalidArtifactException, ArtifactNotFoundException, IOException {
+
+    // get parent artifacts
+    Set<ArtifactRange> parentArtifactRanges = artifactDetail.getMeta().getUsableBy();
+    if (parentArtifactRanges.isEmpty()) {
+      throw new IllegalArgumentException("No parent artifact found for the plugin artifact");
+    }
+
+    // just pick the first parent artifact from the set.
+    ArtifactRange parentArtifactRange = parentArtifactRanges.iterator().next();
+    List<ArtifactDetail> artifactDetails = artifactStore.getArtifacts(parentArtifactRange);
+    if (artifactDetails.isEmpty()) {
+      // should not happen
+      throw new InvalidArtifactException(String.format("No Artifact found in the range %s", parentArtifactRange));
+    }
+
+    // return the first one from the artifact details list.
+    return artifactDetails.get(0).getDescriptor();
+  }
+
+  private String invokePluginEndpoint(Id.Namespace namespace, ArtifactDetail artifactDetail, String pluginType,
+                                      String pluginName, ArtifactDescriptor parentArtifactDescriptor,
+                                      String methodName, String requestBody)
+    throws NotFoundException, IOException, BadRequestException, ClassNotFoundException, InvocationTargetException {
+
+    try (CloseableClassLoader parentClassLoader =
+           artifactClassLoaderFactory.createClassLoader(parentArtifactDescriptor.getLocation())) {
+      Id.Artifact artifactId = Id.Artifact.from(namespace, artifactDetail.getDescriptor().getArtifactId());
+      File stageDir = DirUtils.createTempDir(baseUnpackDir);
+
+      try (PluginInstantiator pluginInstantiator = new PluginInstantiator(cConf, parentClassLoader, stageDir)) {
```

Comment:
please open a jira and add a todo for caching instantiators and address after this.  The way it is now, every call is incredibly expensive.  It will create a new instantiator, new local directory, pull parent artifact, pull plugin artifact every single call.  There needs to be much better handling of this.